runtime/block_reduction.cu

```diff
@@ -111,3 +111,154 @@ __device__ void blockReduce(
+  static_assert(
+      N == 2 || N == 4 || N == 8 || N == 16,
+      "N should be a valid vectorization factor, one of (2, 4, 8, 16)!");
+
+  bool should_write =
+      index_utils::maskedIsZero<X_REDUCE, Y_REDUCE, Z_REDUCE>(threadIdx);
+
+  unsigned int reduction_size =
+      index_utils::maskedSize<X_REDUCE, Y_REDUCE, Z_REDUCE>(blockDim);
+
+  unsigned int reduction_tid =
+      index_utils::maskedOffset<X_REDUCE, Y_REDUCE, Z_REDUCE>(
+          threadIdx, blockDim);
+
+  unsigned int reduction_idx =
+      index_utils::maskedOffset<!X_REDUCE, !Y_REDUCE, !Z_REDUCE>(
+          threadIdx, blockDim);
+
+  // Adjust shared memory offset for array processing
+  unsigned int smem_offset =
+      (reduction_idx * reduction_size + reduction_tid) * N;
+  if (read_pred) {
+    // This section calculates the number of vectorized load operations required
+    // to fetch all elements of an array into shared memory, assuming each load
+    // can transfer up to 16 bytes. For example, with fusion input vectorized by
+    // 8 (N = 8) and computations in fp32 (sizeof(T) = 4 bytes), the total data
+    // size is 4 * 8 = 32 bytes, necessitating 32 / 16 = 2 load transactions.
+    // Each transaction loads 16 / 4 (bytes per element) = 4 elements.
+    if constexpr (sizeof(T) * N <= 16) {
+      loadGeneric<T, N>(shared_mem + smem_offset, const_cast<T*>(inp_val));
+    } else {
+      constexpr unsigned int total_loads = sizeof(T) * N / 16;
+      constexpr unsigned int elements_per_load = 16 / sizeof(T);
+      static_assert(
+          sizeof(T) * N == 16 * total_loads,
+          "This combination of vectorization factor and data type is not supported!");
+      static_assert(
+          sizeof(T) * elements_per_load == 16,
+          "This data type is not supported!");
+#pragma unroll
+      for (unsigned int i = 0; i < total_loads; ++i) {
+        loadGeneric<T, elements_per_load>(
+            shared_mem + smem_offset + i * elements_per_load,
+            const_cast<T*>(inp_val) + i * elements_per_load);
+      }
+    }
+  } else {
+#pragma unroll
+    for (int i = 0; i < N; ++i) {
+      shared_mem[smem_offset + i] = init_val;
+    }
+  }
+
+  block_sync::sync<Aligned>();
+
+  // Reduce down to nearest power of 2 for the tree reduction:
+  int np2 = 1 << (31 - __clz(reduction_size));
+
+  // Perform parallel reduction for each element in the array
+  if (reduction_tid < np2 && reduction_tid + np2 < reduction_size) {
+#pragma unroll
+    for (int i = 0; i < N; ++i) {
+      reduction_op(
+          shared_mem[smem_offset + i], shared_mem[smem_offset + np2 * N + i]);
+    }
+  }
+
+  block_sync::sync<Aligned>();
+
+  for (int factor = np2 / 2; factor > 1; factor >>= 1) {
+    if (reduction_tid < factor) {
+#pragma unroll
+      for (int i = 0; i < N; ++i) {
+        reduction_op(
+            shared_mem[smem_offset + i],
+            shared_mem[smem_offset + factor * N + i]);
+      }
+    }
+    block_sync::sync<Aligned>();
+  }
```

Comment:
Suppose `N=8` and the number of threads per thread block is 256, i.e., 8 warps. Once all the initial inputs are uploaded to the shared memory, the rest can be just done as warp reductions by letting each warp take care of one of `N` reductions. That would eliminate all the synchreads. Furthermore, unlike this implementation, the ratio of threads being idle is much smaller.

That's how the grouped welford is implemented. That was important in the case of welford particularly because it has three values to reduce. I'm not sure if it's also the case here, but I suppose that's true to some extent.

Reply:
Sounds reasonable. Let me do somes tests of using warp reduction. Just curious, if it is more efficient why the original block reduction is not doing this?

Reply:
Because `N` is just 1, there's not much we can do. I tried the warp reduction once `factor` becomes 32. It was long time ago, but the difference was negligible.

Reply:
And before trying to implement it, let's gather more data. I'm pretty sure if it's more efficient, but I don't know if it's necessary.